pkg/util/keymutex/keymutex_test.go

```diff
@@ -65,6 +68,50 @@ func Test_DoubleLock_DoubleUnlock(t *testing.T) {
 	km.UnlockKey(key)
 	verifyCallbackHappens(t, callbackCh2ndLock)
 	km.UnlockKey(key)
+	verifyMutexesUsed(t, km.(*keyMutex), 0)
+}
+
+func benchmarkKeyMutex(b *testing.B, numKeys int, work bool) {
+	km := NewKeyMutex()
+	keys := make([]string, numKeys)
+	if numKeys <= 0 {
+		return
+	}
+	for i := 0; i < numKeys; i++ {
+		keys[i] = fmt.Sprintf("fakeid-%d", i)
+	}
+	b.RunParallel(func(pb *testing.PB) {
+		foo := 0
+		keyid := 0
+		for pb.Next() {
+			key := keys[keyid%numKeys]
+			keyid++
+			km.LockKey(key)
+			if work {
+				for i := 0; i < 100; i++ {
```

Comment:
Realistically, I think the work being done by users of this library is much more expensive than this (making system/shell calls).  I would expect that as the amount of work increases, then the overhead of refcnt checking on Unlock() becomes negligible.

Reply:
Updated.

Before:

```
$ go test -bench . k8s.io/kubernetes/pkg/util/keymutex
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/util/keymutex
BenchmarkKeyMutex/Key1Work0-6         	 3000000	       403 ns/op
BenchmarkKeyMutex/Key1Work100-6       	  300000	      4329 ns/op
BenchmarkKeyMutex/Key1Work10000-6     	      50	  30296084 ns/op
BenchmarkKeyMutex/Key10Work0-6        	 3000000	       461 ns/op
BenchmarkKeyMutex/Key10Work100-6      	  500000	      2582 ns/op
BenchmarkKeyMutex/Key10Work10000-6    	     200	   9437800 ns/op
PASS
ok  	k8s.io/kubernetes/pkg/util/keymutex	11.201s
```

After:

```
$ go test -bench . k8s.io/kubernetes/pkg/util/keymutex
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/util/keymutex
BenchmarkKeyMutex/Key1Work0-6         	 3000000	       435 ns/op
BenchmarkKeyMutex/Key1Work100-6       	  300000	      5386 ns/op
BenchmarkKeyMutex/Key1Work10000-6     	      50	  32181779 ns/op
BenchmarkKeyMutex/Key10Work0-6        	 3000000	       536 ns/op
BenchmarkKeyMutex/Key10Work100-6      	  500000	      3419 ns/op
BenchmarkKeyMutex/Key10Work10000-6    	     200	   9587626 ns/op
PASS
ok  	k8s.io/kubernetes/pkg/util/keymutex	12.525s
```
`benchcmp` result:

```
$ benchcmp 1.txt 2.txt 
benchmark                              old ns/op     new ns/op     delta
BenchmarkKeyMutex/Key1Work0-6          403           435           +7.94%
BenchmarkKeyMutex/Key1Work100-6        4329          5386          +24.42%
BenchmarkKeyMutex/Key1Work10000-6      30296084      32181779      +6.22%
BenchmarkKeyMutex/Key10Work0-6         461           536           +16.27%
BenchmarkKeyMutex/Key10Work100-6       2582          3419          +32.42%
BenchmarkKeyMutex/Key10Work10000-6     9437800       9587626       +1.59%
```

@msau42 Yes, the overhead of refcnt checking becomes negligible as the amount of work increases.